## Learning objectives

read and load selected columns from a large CSV with pandas and define column data types

convert pandas columns into NumPy arrays for numerical analysis

apply NumPy methods such as mean(), nan_to_num(), to_numpy(), and array broadcasting

calculate summary statistics (min, max, average) using NumPy arrays

compare performance of pandas vs NumPy for filtering and aggregation tasks

measure execution time of operations to evaluate efficiency

practice using NumPy for high-performance computing in data analysis

In [54]:
import pandas as pd

In [55]:
pd.set_option("display.max_column", None)

## Question 1

In [6]:
# Part a
cols = [
    'DepDelay', 'ArrDelay', 'Distance',
    'CarrierDelay', 'WeatherDelay',
    'DepTime', 'ArrTime', 'Diverted', 'AirTime'
]

In [7]:
col_types = {
    'DepDelay': 'float64',
    'ArrDelay': 'float64',
    'Distance': 'float64',
    'CarrierDelay': 'float64',
    'WeatherDelay': 'float64',
    'DepTime': 'float64',
    'ArrTime': 'float64',
    'Diverted': 'int64',
    'AirTime': 'float64'
}

In [8]:
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/2014.csv", usecols = cols, dtype = col_types)

In [9]:
myDF.head()

DepTime  DepDelay  ArrTime  ArrDelay  Diverted  AirTime  Distance  \
0    935.0      -5.0   1051.0      -4.0         0     56.0     328.0   
1    951.0      11.0   1115.0      20.0         0     54.0     328.0   
2   1144.0       9.0   1302.0       2.0         0     57.0     328.0   
3   1134.0      -1.0   1253.0      -7.0         0     53.0     328.0   
4   1129.0      -6.0   1244.0     -16.0         0     52.0     328.0   

   CarrierDelay  WeatherDelay  
0           NaN           NaN  
1          11.0           0.0  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN

Markdown notes and sentences and analysis written here.

## Question 2

In [11]:
import numpy as np

In [12]:
# Part a
myDelays = myDF['DepDelay'].to_numpy()

In [15]:
# Part b
print("The shape of my numpy array is",myDelays.shape," and the type is:", myDelays.dtype) 

The shape of my numpy array is (5819811,)  and the type is: float64


In [20]:
# Part c
my_clean_myDelays = np.nan_to_num(myDelays, nan = 0)

In [21]:
# Part d
myLargerValue = my_clean_myDelays + 15

In [30]:
before = np.mean(my_clean_myDelays)

In [29]:
after = np.mean(myLargerValue)

In [36]:
print("The avergae Departure Delay before adding 15 minutes is:",(before),"\n","The average Departure Delay after adding 15 minutes is:",(after))

The avergae Departure Delay before adding 15 minutes is: 10.417677824932802 
 The average Departure Delay after adding 15 minutes is: 25.417677824932802


Markdown notes and sentences and analysis written here.

## Question 3

In [41]:
# Part a
Max = np.max(my_clean_myDelays)

In [42]:
Min = np.min(my_clean_myDelays)

In [48]:
print("Max Arrival Delay:",(Max),"mintues\n","Min Arrival Delay:",(Min),"minutes")

Max Arrival Delay: 2402.0 mintues
 Min Arrival Delay: -251.0 minutes


Markdown notes and sentences and analysis written here.

## Question 4

In [49]:
import time

In [58]:
# Part a b c 
start_time = time.time()

delayed_flights = myDF[(myDF['DepDelay'] > 60) | (myDF['ArrDelay'] > 60)]

my_average_overall_time = delayed_flights['Distance'].mean()

print(f"The average overall score for flights is {my_average_overall_time}")

end_time = time.time()
print(f"Used time is {end_time - start_time}")

The average overall score for flights is 793.1002773092284
Used time is 0.032134294509887695


Markdown notes and sentences and analysis written here.

## Question 5

In [59]:
# Part a b c 
start_time = time.time()
myDepDelayArray = myDF['DepDelay'].to_numpy()
myArrDelayArray = myDF['ArrDelay'].to_numpy()
myOverallArray = myDF['Distance'].to_numpy()

my_Filtered_overall_Scores = myOverallArray[(myDepDelayArray > 60) | (myArrDelayArray > 60)]

my_average_overall_score_numpy = np.mean(my_Filtered_overall_Scores)

print(f"The average overall score for flights is {my_average_overall_score_numpy}")

end_time = time.time()
print(f"Used time is {end_time - start_time}")

The average overall score for flights is 793.1002773092284
Used time is 0.01387476921081543


Part D:
This program takes 0.0139 to run the program while the other one takes 0.0321 seconds to run therefore making Numpy the faster runtime code

Part E: 
As stated in Part D, Numpy is around two times faster to run then Pandas which means numpy is faster and more efficient. This makes me believe that Numpy is more for math since all of these were about finding math while Pandas is maybe for finding more data.